File: Data Structures/linked_list/swift/LinkedList.swift
import Foundation

print("--- Linked List ---")
let linkedList = LinkedList<Int>()

linkedList.add(1)
linkedList.add(2)
linkedList.add(3)

print(linkedList) // [1, 2, 3]

if let removed = try? linkedList.remove() { print(removed) } // 3
if let removed = try? linkedList.remove() { print(removed) } // 2
if let removed = try? linkedList.remove() { print(removed) } // 1

print(linkedList.isEmpty) // true
print(linkedList) // []

linkedList.add(1)
linkedList.add(2)

if let peeked = try? linkedList.peek() { print(peeked) } // 2

print(linkedList.isEmpty) // false
print(linkedList) // [1, 2]


// --- Implementations ---

class Node<T> {
    let value: T
    var next: Node<T>?

    init(value: T) {
        self.value = value
    }
}

class LinkedList<T>: CustomStringConvertible {
    var head: Node<T>?

    // Add an element to the linked list.
    func add(_ value: T) {
        let node = Node(value: value)

        // If the head is nil, assign the node to the head.
        if head == nil {
            head = node
            return
        }

        // If the head is not nil, traverse to the end.
        var currentNode = head!
        while currentNode.next != nil {
            currentNode = currentNode.next!
        }
        currentNode.next = node
    }

    // Add all elements to the linked list.
    func addAll(_ values: [T]) {
        for value in values {
            add(value)
        }
    }

    // Remove the last element from the linked list and return it.
    func remove() throws -> T {
        guard let headNode = head else {
            throw LinkedListError.empty
        }

        if headNode.next == nil {
            let value = headNode.value
            head = nil
            return value
        }

        var currentNode = headNode
        var previousNode = headNode

        while currentNode.next != nil {
            previousNode = currentNode
            currentNode = currentNode.next!
        }

        previousNode.next = nil
        return currentNode.value
    }

    // Get the last element of the linked list without removing it.
    func peek() throws -> T {
        guard let headNode = head else { throw LinkedListError.empty }

        var currentNode = headNode
        while currentNode.next != nil {
            currentNode = currentNode.next!
        }
        return currentNode.value
    }

    var isEmpty: Bool {
        return head == nil
    }

    var description: String {
        var list: [T] = []
        var currentNode = head
        while let node = currentNode {
            list.append(node.value)
            currentNode = node.next
        }
        return "\(list)"
    }
}

class LinkedListWithTail<T> {
    var head: Node<T>?
    var tail: Node<T>?

    func add(_ value: T) {
        let node = Node(value: value)
        if head == nil {
            head = node
            tail = node
            return
        }
        tail?.next = node
        tail = node
    }
}

enum LinkedListError: Error {
    case empty
}

File: Data Structures/linked_list/dart/linked_list.dart
void main() {
  final linkedList = LinkedList<int>();

  linkedList.add(1);
  linkedList.add(2);
  linkedList.add(3);

  print(linkedList); // [1, 2, 3]

  print(linkedList.remove()); // 3
  print(linkedList.remove()); // 2
  print(linkedList.remove()); // 1

  print(linkedList.isEmpty); // true
  print(linkedList); // []

  linkedList.add(1);
  linkedList.add(2);

  print(linkedList.peek()); // 2

  print(linkedList.isEmpty); // false
  print(linkedList); // [1, 2]
}

class Node<T> {
  final T value;
  Node<T>? next;

  Node(this.value);
}

class LinkedList<T> {
  Node<T>? head;

  /// Add an element to the linked list.
  void add(T value) {
    final node = Node(value);

    /// If the head is null, assign the node to the head.
    if (head == null) {
      head = node;
      return;
    }

    /// If the head is not null, reference the node to the next node.
    Node currentNode = head!;
    // Iterate through head until the reference to the next node is null.
    while (currentNode.next != null) {
      currentNode = currentNode.next!;
    }
    currentNode.next = node;
  }

  /// Add all elements to the linked list.
  void addAll(Iterable<T> values) {
    for (final value in values) {
      add(value);
    }
  }

  /// Remove the last element from the linked list and return it.
  T remove() {
    if (head == null) {
      throw StateError('Linked list is empty');
    } else if (head!.next == null) {
      final value = head!.value;
      head = null;
      return value;
    }

    /// If the head is not null, reference the node to the next node.
    Node currentNode = head!;
    Node previousNode = head!;

    // Iterate through head until the reference to the next node is null.
    while (currentNode.next != null) {
      previousNode = currentNode;
      currentNode = currentNode.next!;
    }

    previousNode.next = null;
    return currentNode.value;
  }

  /// Get the last element of the linked list without removing it.
  T peek() {
    if (head == null) throw StateError('Linked list is empty');

    /// If the head is not null, reference the node to the next node.
    Node currentNode = head!;

    // Iterate through head until the reference to the next node is null.
    while (currentNode.next != null) {
      currentNode = currentNode.next!;
    }

    return currentNode.value;
  }

  bool get isEmpty => head == null;

  int get length {
    if (head == null) return 0;

    int length = 1;
    Node currentNode = head!;

    // Iterate through head until the reference to the next node is null.
    while (currentNode.next != null) {
      currentNode = currentNode.next!;
      length++;
    }

    return length;
  }

  @override
  String toString() {
    if (head == null) return '[]';

    final List<T> list = [];
    Node currentNode = head!;

    // Iterate through head until the reference to the next node is null.
    while (currentNode.next != null) {
      list.add(currentNode.value);
      currentNode = currentNode.next!;
    }

    list.add(currentNode.value);
    return list.toString();
  }
}

class LinkedListWithTail<T> {
  Node<T>? head;
  Node<T>? tail; // Added tail pointer

  /// Add an element to the linked list.
  /// To Understand this code, you need to understand the memory allocation of variables in Dart.
  void add(T value) {
    final node = Node(value);

    if (head == null) {
      // If the list is empty, set both head and tail to the new node.
      head = node;
      tail = node; // Update tail to point to the new node
      return;
    }

    // If the list is not empty, append the new node after the tail.
    // First addition, A location for the Node, Head, Tail = A location for Node
    // Next addition, B location, Head/Tail(A).next = B location, Tail = B location
    // Next addition, C location, Tail(B).next = C location, Tail = C location
    // Result Head -> A -> B -> C -> Tail
    tail!.next = node;
    tail = node; // Update tail to point to the new node
  }

  // Other methods remain the same...
}
